cylc/flow/ws_data_mgr.py

```diff
@@ -221,138 +312,250 @@ def _generate_ghost_families(self, family_proxies=None, cycle_points=None):
+        self.pool_points = set(list(self.schd.pool.pool))
+        # No action if pool is not yet initiated.
+        if not self.pool_points:
+            return
+        # Increment edges:
+        # - Initially for each cycle point in the pool.
+        # - For each new cycle point thereafter.
+        # Using difference and pointwise allows for historical
+        # task insertion (in gaps).
+        for point in self.pool_points.difference(old_pool_points):
+            # All family & task cycle instances are generated and
+            # populated with static data as 'ghost nodes'.
+            self.generate_graph_elements(
+                self.edges, self.task_proxies, self.family_proxies,
+                point, point)
+            has_updated = True
+        # Prune data store by cycle point for:
+        # - Edge target cycle points are not in the pool
+        #   and source cycle points are not greater than pool max.
+        # - Edge source cycle points where neither itself and
+        #   all it's corresponding target cycle points are not in the pool.
+        # This ensures a buffer of sources and targets in front and behind the
+        # task pool, while accommodating exceptions such as ICP dependencies.
+        # TODO: Turn nodes back to ghost if not in pool? (for suicide)
+        prune_points = set()
+        max_point = max(self.pool_points)
+        for s_point, t_points in list(self.edge_points.items()):
+            if s_point <= max_point:
+                if (s_point not in self.pool_points and
+                        t_points.isdisjoint(self.pool_points)):
+                    prune_points.add(str(s_point))
+                    prune_points.union((str(t_p) for t_p in t_points))
+                    del self.edge_points[s_point]
+                    continue
+                t_diffs = t_points.difference(self.pool_points)
+                if t_diffs:
+                    prune_points.union((str(t_p) for t_p in t_diffs))
+                    self.edge_points[s_point].difference_update(t_diffs)
+        # Action pruning if any eligible cycle points are found.
+        if prune_points:
+            self.prune_points(prune_points)
+            has_updated = True
+        if has_updated:
+            self.update_workflow()
+
+    def initiate_data_model(self, reload=False):
+        """Initiate or Update data model on start/restart/reload."""
+        # Reset attributes on reload:
+        if reload:
+            self.pool_points = set()
+            self.edge_points = {}
+            self.all_states = []
+            self.cycle_states = {}
+            self.state_count_cycles = {}
+            self.task_proxies = {}
+            self.family_proxies = {}
+            self.edges = {}
+            self.graph = PbEdges()
+        # Static elements
+        self.generate_definition_elements()
+        self.increment_graph_elements()
+
+    def update_task_proxies(self, updated_tasks=None):
```

Comment:
I've noticed `updated_tasks` here can have duplicate elements. I think due to the same task instance having "changed state" and "changed pools" (runhead to live).  May have some impact on efficiency as the code below isn't checking to see if we've already processed each one in the loop.

Reply:
It would be good to have in the docstrings exactly what incoming data structures are supposed to contain as the names aren't always sufficiently descriptive. (`updated_tasks` is list of task TaskProxy instances that have changed state OR been released from the runhead pool?).